File: windows/PabloCompanion/ViewModels/PracticeViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using PabloCompanion.Models;
using PabloCompanion.Services;

namespace PabloCompanion.ViewModels;

/// <summary>
/// Orchestrates a practice session: topic selection → WebSocket connection →
/// mic capture → audio playback → session teardown.
/// </summary>
public sealed partial class PracticeViewModel : ObservableObject, IDisposable
{
    // ── State ───────────────────────────────────────────────────────────────

    public enum Phase { Idle, LoadingTopics, PickingTopic, Connecting, Active, Ending, Ended, Error }

    [ObservableProperty]
    public partial Phase CurrentPhase { get; set; } = Phase.Idle;

    [ObservableProperty]
    public partial PracticeTopic[] Topics { get; set; } = [];

    [ObservableProperty]
    public partial PracticeTopic? SelectedTopic { get; set; }

    [ObservableProperty]
    public partial string? SessionId { get; set; }

    [ObservableProperty]
    public partial int DurationSeconds { get; set; }

    [ObservableProperty]
    public partial float MicLevel { get; set; }

    [ObservableProperty]
    public partial float PabloLevel { get; set; }

    [ObservableProperty]
    public partial PracticeWebSocketClient.PabloState PabloState { get; set; } = PracticeWebSocketClient.PabloState.Listening;

    [ObservableProperty]
    public partial string? ErrorMessage { get; set; }

    [ObservableProperty]
    public partial int EndedDurationSeconds { get; set; }

    // ── Dependencies ────────────────────────────────────────────────────────

    private readonly PracticeApiClient _apiClient;
    private readonly PracticeWebSocketClient _wsClient = new();
    private readonly PracticeMicCapture _micCapture = new();
    private readonly PracticeAudioPlayer _audioPlayer = new();

    private CancellationTokenSource? _durationCts;
    private CancellationTokenSource? _reconnectCts;
    private DateTime? _sessionStartTime;
    private int _reconnectAttempts;
    private const int MaxReconnectAttempts = 3;

    /// <summary>
    /// The mic device ID to use (from RecordingViewModel's selected device).
    /// </summary>
    public string? SelectedMicId { get; set; }

    public bool IsSessionActive => CurrentPhase is Phase.Active or Phase.Ending or Phase.Connecting;

    public PracticeViewModel(PracticeApiClient apiClient)
    {
        _apiClient = apiClient;
        ConfigureCallbacks();
    }

    // ── Topic loading ───────────────────────────────────────────────────────

    [RelayCommand]
    private async Task LoadTopicsAsync()
    {
        CurrentPhase = Phase.LoadingTopics;
        try
        {
            Topics = await _apiClient.FetchTopicsAsync();
            CurrentPhase = Phase.PickingTopic;
        }
        catch (Exception ex)
        {
            ErrorMessage = $"Failed to load practice topics: {ex.Message}";
            CurrentPhase = Phase.Idle;
        }
    }

    // ── Session lifecycle ───────────────────────────────────────────────────

    [RelayCommand]
    private async Task StartSessionAsync(PracticeTopic topic)
    {
        App.Log($"[Practice] StartSessionAsync topic={topic.Id} ({topic.Name})");
        SelectedTopic = topic;
        CurrentPhase = Phase.Connecting;

        try
        {
            // 1. Create session via REST
            App.Log($"[Practice] POST /api/practice/sessions baseUrl={_apiClient.BaseUrl}");
            var response = await _apiClient.CreateSessionAsync(topic.Id);
            SessionId = response.SessionId;
            App.Log($"[Practice] CreateSession ok session={response.SessionId} ticket.len={response.WsTicket.Length}");

            // 2. Connect WebSocket
            var wsUri = _apiClient.BuildWebSocketUri(response.WsTicket);
            App.Log($"[Practice] ws connect → {wsUri.Host}:{wsUri.Port}{wsUri.AbsolutePath}");
            await _wsClient.ConnectAsync(wsUri);
            App.Log($"[Practice] ws connect returned, state={_wsClient.State}");

            // 3. Start mic capture
            _micCapture.Start(SelectedMicId);
            App.Log($"[Practice] mic started (deviceId={SelectedMicId ?? "default"})");

            // 4. Start audio player
            _audioPlayer.Start();
            App.Log("[Practice] audio player started");

            // 5. Wait briefly for auth, then send session_start
            await Task.Delay(500);
            _wsClient.StartSession(response.SessionId);
            App.Log($"[Practice] sent session_start session={response.SessionId}");
        }
        catch (Exception ex)
        {
            App.LogException("[Practice] StartSessionAsync failed", ex);
            Cleanup();
            ErrorMessage = $"Failed to start practice session: {ex.Message}";
            CurrentPhase = Phase.Idle;
        }
    }

    [RelayCommand]
    private void EndSession()
    {
        if (CurrentPhase != Phase.Active) return;
        CurrentPhase = Phase.Ending;
        _wsClient.EndSession();
    }

    [RelayCommand]
    private void PauseAudio() => _wsClient.PauseAudio();

    [RelayCommand]
    private void ResumeAudio() => _wsClient.ResumeAudio();

    [RelayCommand]
    private void Dismiss()
    {
        Cleanup();
        CurrentPhase = Phase.Idle;
        SelectedTopic = null;
        SessionId = null;
        DurationSeconds = 0;
    }

    // ── Callbacks ───────────────────────────────────────────────────────────

    private void ConfigureCallbacks()
    {
        _wsClient.ConnectionStateChanged += OnConnectionStateChanged;
        _wsClient.PabloStateChanged += state => DispatcherQueue(() => PabloState = state);
        _wsClient.AudioReceived += OnAudioReceived;
        _wsClient.SessionStarted += OnSessionStarted;
        _wsClient.SessionEnded += OnSessionEnded;
        _wsClient.ErrorOccurred += OnError;

        _micCapture.AudioFrameReady += frame => _wsClient.SendAudioFrame(frame);
        _micCapture.LevelUpdated += level => DispatcherQueue(() => MicLevel = level);
        _audioPlayer.LevelUpdated += level => DispatcherQueue(() => PabloLevel = level);
    }

    private void OnConnectionStateChanged(PracticeWebSocketClient.ConnectionState state)
    {
        DispatcherQueue(() =>
        {
            if (state == PracticeWebSocketClient.ConnectionState.Disconnected && CurrentPhase == Phase.Active)
            {
                AttemptReconnection();
            }
            else if (state == PracticeWebSocketClient.ConnectionState.Active)
            {
                _reconnectAttempts = 0;
            }
        });
    }

    private void OnAudioReceived(byte[] pcmData, bool isFinal)
    {
        _audioPlayer.Enqueue(pcmData);
        if (isFinal)
        {
            DispatcherQueue(() => PabloState = PracticeWebSocketClient.PabloState.Listening);
        }
    }

    private void OnSessionStarted(string sessionId)
    {
        DispatcherQueue(() =>
        {
            SessionId = sessionId;
            CurrentPhase = Phase.Active;
            _sessionStartTime = DateTime.UtcNow;
            StartDurationTimer();
        });
    }

    private void OnSessionEnded(int durationSeconds)
    {
        DispatcherQueue(() =>
        {
            Cleanup();
            EndedDurationSeconds = durationSeconds;
            CurrentPhase = Phase.Ended;
        });
    }

    private void OnError(string message, bool fatal)
    {
        DispatcherQueue(() =>
        {
            if (fatal)
            {
                Cleanup();
                CurrentPhase = Phase.Error;
            }
            ErrorMessage = message;
        });
    }

    // ── Reconnection ────────────────────────────────────────────────────────

    private void AttemptReconnection()
    {
        if (SessionId == null || _reconnectAttempts >= MaxReconnectAttempts)
        {
            Cleanup();
            ErrorMessage = "Connection lost";
            CurrentPhase = Phase.Error;
            return;
        }

        _reconnectAttempts++;
        CurrentPhase = Phase.Connecting;

        var lastSeq = _wsClient.LastReceivedSequence;
        var sessionId = SessionId!;

        _reconnectCts?.Cancel();
        _reconnectCts = new CancellationTokenSource();
        var ct = _reconnectCts.Token;

        _ = Task.Run(async () =>
        {
            // Exponential backoff: 0.5s, 1s, 2s
            var delay = (int)(500 * Math.Pow(2, _reconnectAttempts - 1));
            await Task.Delay(delay, ct);
            if (ct.IsCancellationRequested) return;

            try
            {
                var ticket = await _apiClient.FetchTicketAsync();
                var wsUri = _apiClient.BuildWebSocketUri(ticket);

                _wsClient.Disconnect(forReconnect: true);
                await _wsClient.ConnectAsync(wsUri);

                await Task.Delay(500, ct);
                if (ct.IsCancellationRequested) return;

                _wsClient.ResumeSession(sessionId, lastSeq);
            }
            catch
            {
                if (!ct.IsCancellationRequested)
                {
                    DispatcherQueue(AttemptReconnection);
                }
            }
        }, ct);
    }

    // ── Duration timer ──────────────────────────────────────────────────────

    private void StartDurationTimer()
    {
        _durationCts?.Cancel();
        _durationCts = new CancellationTokenSource();
        var ct = _durationCts.Token;

        _ = Task.Run(async () =>
        {
            while (!ct.IsCancellationRequested)
            {
                await Task.Delay(1000, ct);
                if (ct.IsCancellationRequested || _sessionStartTime == null) break;
                var elapsed = (int)(DateTime.UtcNow - _sessionStartTime.Value).TotalSeconds;
                DispatcherQueue(() => DurationSeconds = elapsed);
            }
        }, ct);
    }

    // ── Cleanup ─────────────────────────────────────────────────────────────

    private void Cleanup()
    {
        _reconnectCts?.Cancel();
        _reconnectCts = null;
        _reconnectAttempts = 0;
        _durationCts?.Cancel();
        _durationCts = null;
        _micCapture.Stop();
        _audioPlayer.Stop();
        _wsClient.Disconnect();
        MicLevel = 0;
        PabloLevel = 0;
        PabloState = PracticeWebSocketClient.PabloState.Listening;
    }

    public void Dispose()
    {
        Cleanup();
        _wsClient.Dispose();
        _micCapture.Dispose();
        _audioPlayer.Dispose();
    }

    // ── Dispatcher helper ───────────────────────────────────────────────────

    private static void DispatcherQueue(Action action)
    {
        if (Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread() != null)
        {
            action();
        }
        else
        {
            App.UiDispatcherQueue?.TryEnqueue(() => action());
        }
    }
}

File: mac/PabloCompanion/ViewModels/PracticeViewModel.swift
import Foundation
import os
import PracticeClientCore

/// Orchestrates a practice session: topic selection → WebSocket connection →
/// mic capture → audio playback → session teardown.
@MainActor
@Observable
final class PracticeViewModel {
    // MARK: - State

    enum Phase: Equatable {
        case idle
        case loadingTopics
        case pickingTopic
        case connecting
        case active
        case ending
        case ended(durationSeconds: Int)
        case error(String)
    }

    var phase: Phase = .idle
    var topics: [PracticeTopic] = []
    var selectedTopic: PracticeTopic?
    var sessionId: String?
    var duration: TimeInterval = 0
    var micLevel: Float = 0
    var pabloLevel: Float = 0
    var pabloState: PracticeWebSocketClient.PabloState = .listening
    var errorMessage: String?
    var showError = false

    // MARK: - Dependencies

    var backendURL = AppConstants.defaultBackendAPIURL {
        didSet { apiClient.baseURL = backendURL }
    }

    private let apiClient = PracticeAPIClient()
    private let wsClient = PracticeWebSocketClient()
    private let micCapture = PracticeMicCapture()
    private let audioPlayer = PracticeAudioPlayer()
    private let logger = Logger(subsystem: AppConstants.appBundleID, category: "PracticeViewModel")
    private var durationTimer: Task<Void, Never>?
    private var reconnectTask: Task<Void, Never>?
    private var sessionStartTime: Date?
    private var reconnectAttempts = 0
    private static let maxReconnectAttempts = 3

    /// The mic device UID to use (mirrors RecordingViewModel.selectedMicID).
    var selectedMicID: String?

    init() {
        // Preserve the app's real client-type header; the core defaults to a dev value.
        apiClient.clientType = "pablo-companion-macos/\(AppConstants.appVersion)"
        configureCallbacks()
    }

    // MARK: - Auth

    func configureAuth(
        getToken: @escaping @Sendable () async throws -> String,
        onAuthRejected: ((Bool) -> Void)? = nil
    ) {
        apiClient.getToken = getToken
        // The practice client has no error envelope; a 401 here is a plain
        // server-side session rejection (never a parsed idle-timeout code).
        apiClient.onUnauthorized = onAuthRejected.map { handler in { handler(false) } }
    }

    // MARK: - Topic loading

    func loadTopics() async {
        phase = .loadingTopics
        do {
            topics = try await apiClient.fetchTopics()
            phase = .pickingTopic
        } catch {
            logger.error("Failed to load topics: \(error.localizedDescription)")
            showErrorAlert("Failed to load practice topics: \(error.localizedDescription)")
            phase = .idle
        }
    }

    // MARK: - Session lifecycle

    func startSession(topic: PracticeTopic) async {
        selectedTopic = topic
        phase = .connecting

        do {
            // 1. Create session via REST
            let response = try await apiClient.createSession(topicId: topic.id)
            sessionId = response.sessionId
            logger.info("Practice session created: \(response.sessionId)")

            // 2. Connect WebSocket using the single-use ticket from session creation
            guard let wsURL = apiClient.webSocketURL(ticket: response.wsTicket) else {
                throw APIError.invalidResponse
            }
            wsClient.connect(url: wsURL)

            // 3. Start mic capture (will begin sending frames once WS is active)
            try micCapture.start(micDeviceID: selectedMicID)

            // 4. Start audio player (for Pablo's responses)
            audioPlayer.start()

            // 5. Wait briefly for auth, then send session_start
            try await Task.sleep(for: .milliseconds(500))
            wsClient.startSession(sessionId: response.sessionId)

        } catch {
            logger.error("Failed to start practice session: \(error.localizedDescription)")
            cleanup()
            showErrorAlert("Failed to start practice session: \(error.localizedDescription)")
            phase = .idle
        }
    }

    func endSession() {
        guard phase == .active else { return }
        phase = .ending
        wsClient.endSession()
    }

    func pauseAudio() {
        wsClient.pauseAudio()
    }

    func resumeAudio() {
        wsClient.resumeAudio()
    }

    func dismiss() {
        cleanup()
        phase = .idle
        selectedTopic = nil
        sessionId = nil
        duration = 0
    }

    /// Whether the practice session is currently in progress (active or ending).
    var isSessionActive: Bool {
        switch phase {
        case .active, .ending, .connecting:
            true
        default:
            false
        }
    }

    // MARK: - Callbacks

    private func configureCallbacks() {
        configureWebSocketCallbacks()
        configureAudioCallbacks()
    }

    private func configureWebSocketCallbacks() {
        wsClient.onConnectionStateChanged = { [weak self] state in
            Task { @MainActor [weak self] in
                self?.handleConnectionState(state)
            }
        }

        wsClient.onPabloStateChanged = { [weak self] state in
            Task { @MainActor [weak self] in
                self?.pabloState = state
            }
        }

        wsClient.onAudioReceived = { [weak self] pcmData, isFinal in
            self?.audioPlayer.enqueue(pcmData)
            if isFinal {
                Task { @MainActor [weak self] in
                    self?.pabloState = .listening
                }
            }
        }

        wsClient.onSessionStarted = { [weak self] sessionId in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.sessionId = sessionId
                self.phase = .active
                self.sessionStartTime = Date()
                self.startDurationTimer()
                self.logger.info("Practice session active")
            }
        }

        wsClient.onSessionEnded = { [weak self] durationSeconds in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.cleanup()
                self.phase = .ended(durationSeconds: durationSeconds)
                self.logger.info("Practice session ended (\(durationSeconds)s)")
            }
        }

        wsClient.onError = { [weak self] message, fatal in
            Task { @MainActor [weak self] in
                guard let self else { return }
                if fatal {
                    self.cleanup()
                    self.phase = .error(message)
                }
                self.showErrorAlert(message)
            }
        }
    }

    private func configureAudioCallbacks() {
        micCapture.onAudioFrame = { [weak self] pcmData in
            self?.wsClient.sendAudioFrame(pcmData)
        }

        micCapture.onLevelUpdate = { [weak self] level in
            Task { @MainActor [weak self] in
                self?.micLevel = level
            }
        }

        audioPlayer.onLevelUpdate = { [weak self] level in
            Task { @MainActor [weak self] in
                self?.pabloLevel = level
            }
        }
    }

    private func handleConnectionState(_ state: PracticeWebSocketClient.ConnectionState) {
        switch state {
        case .disconnected:
            if case .active = phase {
                // Unexpected disconnect — attempt reconnection
                attemptReconnection()
            }
        case .connecting, .authenticating, .waitingForSession:
            break // already in .connecting phase
        case .active:
            reconnectAttempts = 0 // reset on successful connection
        case .ending:
            break // handled by onSessionEnded
        }
    }

    // MARK: - Reconnection

    /// Attempts to reconnect using a fresh single-use ticket.
    /// Retries up to 3 times within the server's 30-second reconnection window.
    private func attemptReconnection() {
        guard let sessionId, reconnectAttempts < Self.maxReconnectAttempts else {
            logger.error("Reconnection failed — max attempts reached")
            cleanup()
            phase = .error("Connection lost")
            return
        }

        reconnectAttempts += 1
        phase = .connecting
        logger.info("Reconnecting (attempt \(self.reconnectAttempts)/\(Self.maxReconnectAttempts))")

        let lastSeq = wsClient.lastReceivedSequence

        reconnectTask?.cancel()
        reconnectTask = Task { [weak self] in
            guard let self else { return }

            // Brief backoff: 0.5s, 1s, 2s
            let delay = 0.5 * pow(2.0, Double(self.reconnectAttempts - 1))
            try? await Task.sleep(for: .seconds(delay))
            guard !Task.isCancelled else { return }

            do {
                let ticket = try await self.apiClient.fetchTicket()
                guard let wsURL = self.apiClient.webSocketURL(ticket: ticket) else {
                    throw APIError.invalidResponse
                }

                self.wsClient.disconnect(forReconnect: true)
                self.wsClient.connect(url: wsURL)

                // Wait for auth, then resume session
                try await Task.sleep(for: .milliseconds(500))
                guard !Task.isCancelled else { return }
                self.wsClient.resumeSession(sessionId: sessionId, lastSequence: lastSeq)

            } catch {
                guard !Task.isCancelled else { return }
                self.logger.error("Reconnect attempt failed: \(error.localizedDescription)")
                // Will trigger another handleConnectionState(.disconnected) → retry
                self.attemptReconnection()
            }
        }
    }

    // MARK: - Duration timer

    private func startDurationTimer() {
        durationTimer?.cancel()
        durationTimer = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(1))
                guard !Task.isCancelled, let self, let start = self.sessionStartTime else { break }
                self.duration = Date().timeIntervalSince(start)
            }
        }
    }

    // MARK: - Cleanup

    private func cleanup() {
        reconnectTask?.cancel()
        reconnectTask = nil
        reconnectAttempts = 0
        durationTimer?.cancel()
        durationTimer = nil
        micCapture.stop()
        audioPlayer.stop()
        wsClient.disconnect()
        micLevel = 0
        pabloLevel = 0
        pabloState = .listening
    }

    // MARK: - Helpers

    private func showErrorAlert(_ message: String) {
        errorMessage = message
        showError = true
    }
}
